typeof statsOptions.colors ...  Cannot read property 'colors' of undefined

<!--
  Issues are so 🔥

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  👉🏽 Need support, advice, or help? Don't open an issue!
  Head to StackOverflow or https://gitter.im/webpack/webpack.
-->

- Operating System:
- Node Version:
- NPM Version:
- webpack Version:
- webpack-dev-middleware Version:

### Expected Behavior

webpack built 4333dbf8aa8a0bbf4aab in 13716ms
C:\Workspace\gd-diy-apply\node_modules\webpack-dev-middleware\dist\utils\setupHooks.js:68
      } else if (typeof statsOptions.colors === 'undefined' || typeof statsOptions === 'string') {
                                     ^

TypeError: Cannot read property 'colors' of undefined
    at C:\Workspace\gd-diy-apply\node_modules\webpack-dev-middleware\dist\utils\setupHooks.js:68:38
    at processTicksAndRejections (internal/process/task_queues.js:75:11)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! aves@6.3.3 start: `node ./config/command/start`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the aves@6.3.3 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\zhenghang\AppData\Roaming\npm-cache\_logs\2021-01-13T06_27_35_802Z-debug.log

### Actual Behavior

<!-- Remove this section if not reporting a bug or modification request. -->

### Code

```js
// webpack.config.js
// If your code blocks are over 20 lines, please paste a link to a gist
// (https://gist.github.com).
```

```js
// additional code, HEY YO remove this block if you don't need it
```

### How Do We Reproduce?

<!--
  Remove this section if not reporting a bug.

  If your webpack config is over 50 lines long, please provide a URL to a repo
  for your beefy 🍖 app that we can use to reproduce.
-->

--- a/src/utils/setupHooks.js
+++ b/src/utils/setupHooks.js
@@ -14,6 +14,25 @@ export default function setupHooks(context) {
     context.stats = undefined;
   }
 
+  const statsForWebpack4 = webpack.Stats && webpack.Stats.presetToOptions;
+
+  function normalizeStatsOptions(statsOptions) {
+    if (statsForWebpack4) {
+      if (typeof statsOptions === 'undefined') {
+        // eslint-disable-next-line no-param-reassign
+        statsOptions = {};
+      } else if (
+        typeof statsOptions === 'boolean' ||
+        typeof statsOptions === 'string'
+      ) {
+        // eslint-disable-next-line no-param-reassign
+        statsOptions = webpack.Stats.presetToOptions(statsOptions);
+      }
+    }
+
+    return statsOptions;
+  }
+
   function done(stats) {
     // We are now on valid state
     // eslint-disable-next-line no-param-reassign
@@ -33,28 +52,14 @@ export default function setupHooks(context) {
       logger.log('Compilation finished');
 
       let statsOptions = compiler.compilers
-        ? {
-            children: compiler.compilers.map((child) =>
-              // eslint-disable-next-line no-undefined
-              child.options ? child.options.stats : undefined
-            ),
-          }
-        : compiler.options
-        ? compiler.options.stats
-        : // eslint-disable-next-line no-undefined
-          undefined;
-
-      const statsForWebpack4 = webpack.Stats && webpack.Stats.presetToOptions;
+        ? { children: compiler.compilers.map((child) => child.options.stats) }
+        : compiler.options.stats;
 
       if (compiler.compilers) {
         statsOptions.children = statsOptions.children.map(
           (childStatsOptions) => {
-            if (statsForWebpack4) {
-              // eslint-disable-next-line no-param-reassign
-              childStatsOptions = webpack.Stats.presetToOptions(
-                childStatsOptions
-              );
-            }
+            // eslint-disable-next-line no-param-reassign
+            childStatsOptions = normalizeStatsOptions(childStatsOptions);
 
             if (typeof childStatsOptions.colors === 'undefined') {
               // eslint-disable-next-line no-param-reassign
@@ -64,15 +69,12 @@ export default function setupHooks(context) {
             return childStatsOptions;
           }
         );
-      } else if (
-        typeof statsOptions.colors === 'undefined' ||
-        typeof statsOptions === 'string'
-      ) {
-        if (statsForWebpack4) {
-          statsOptions = webpack.Stats.presetToOptions(statsOptions);
-        }
+      } else {
+        statsOptions = normalizeStatsOptions(statsOptions);
 
-        statsOptions.colors = Boolean(colorette.options.enabled);
+        if (typeof statsOptions.colors === 'undefined') {
+          statsOptions.colors = Boolean(colorette.options.enabled);
+        }
       }
 
       // TODO webpack@4 doesn't support `{ children: [{ colors: true }, { colors: true }] }` for stats

--- a/test/fixtures/webpack.array.one-error-one-warning-one-no.js
+++ b/test/fixtures/webpack.array.one-error-one-warning-one-no.js
@@ -1,0 +1,67 @@
+'use strict';
+
+const path = require('path');
+
+module.exports = [
+  {
+    mode: 'development',
+    context: path.resolve(__dirname),
+    entry: './broken.js',
+    output: {
+      filename: 'bundle.js',
+      path: path.resolve(__dirname, '../outputs/one-error-one-warning-one-success/js1'),
+      publicPath: '/static-one/',
+    },
+    infrastructureLogging: {
+      level: 'none'
+    },
+    stats: 'normal'
+  },
+  {
+    mode: 'development',
+    context: path.resolve(__dirname),
+    entry: './warning.js',
+    output: {
+      filename: 'bundle.js',
+      path: path.resolve(__dirname, '../outputs/one-error-one-warning-one-success/js2'),
+      publicPath: '/static-two/',
+    },
+    plugins: [
+      {
+        apply(compiler) {
+          compiler.hooks.emit.tapAsync('WarningPlugin', (compilation, done) => {
+            compilation.warnings.push(new Error('Warning'));
+
+            done();
+          })
+        },
+      }
+    ],
+    infrastructureLogging: {
+      level: 'none'
+    },
+    stats: 'normal'
+  },
+  {
+    mode: 'development',
+    context: path.resolve(__dirname),
+    entry: './foo.js',
+    output: {
+      filename: 'bundle.js',
+      path: path.resolve(__dirname, 'js3'),
+      publicPath: '/static-three/',
+    },
+    module: {
+      rules: [
+        {
+          test: /\.(svg|html)$/,
+          loader: 'file-loader',
+          options: { name: '[name].[ext]' },
+        },
+      ],
+    },
+    infrastructureLogging: {
+      level: 'none'
+    },
+  },
+];

--- a/test/fixtures/webpack.array.one-error-one-warning-one-object.js
+++ b/test/fixtures/webpack.array.one-error-one-warning-one-object.js
@@ -1,0 +1,71 @@
+'use strict';
+
+const path = require('path');
+
+module.exports = [
+  {
+    mode: 'development',
+    context: path.resolve(__dirname),
+    entry: './broken.js',
+    output: {
+      filename: 'bundle.js',
+      path: path.resolve(__dirname, '../outputs/one-error-one-warning-one-success/js1'),
+      publicPath: '/static-one/',
+    },
+    infrastructureLogging: {
+      level: 'none'
+    },
+    stats: 'normal'
+  },
+  {
+    mode: 'development',
+    context: path.resolve(__dirname),
+    entry: './warning.js',
+    output: {
+      filename: 'bundle.js',
+      path: path.resolve(__dirname, '../outputs/one-error-one-warning-one-success/js2'),
+      publicPath: '/static-two/',
+    },
+    plugins: [
+      {
+        apply(compiler) {
+          compiler.hooks.emit.tapAsync('WarningPlugin', (compilation, done) => {
+            compilation.warnings.push(new Error('Warning'));
+
+            done();
+          })
+        },
+      }
+    ],
+    infrastructureLogging: {
+      level: 'none'
+    },
+    stats: 'normal'
+  },
+  {
+    mode: 'development',
+    context: path.resolve(__dirname),
+    entry: './foo.js',
+    output: {
+      filename: 'bundle.js',
+      path: path.resolve(__dirname, 'js3'),
+      publicPath: '/static-three/',
+    },
+    module: {
+      rules: [
+        {
+          test: /\.(svg|html)$/,
+          loader: 'file-loader',
+          options: { name: '[name].[ext]' },
+        },
+      ],
+    },
+    infrastructureLogging: {
+      level: 'none'
+    },
+    stats: {
+      all: false,
+      assets: true,
+    }
+  },
+];

--- a/test/fixtures/webpack.no-stats.config.js
+++ b/test/fixtures/webpack.no-stats.config.js
@@ -1,0 +1,25 @@
+'use strict';
+
+const path = require('path');
+
+module.exports = {
+    mode: 'development',
+    context: path.resolve(__dirname),
+    entry: './foo.js',
+    output: {
+        filename: 'bundle.js',
+        path: path.resolve(__dirname, '../outputs/basic'),
+    },
+    module: {
+        rules: [
+            {
+                test: /\.(svg|html)$/,
+                loader: 'file-loader',
+                options: { name: '[name].[ext]' },
+            },
+        ],
+    },
+    infrastructureLogging: {
+        level: 'none'
+    },
+};

--- a/test/logging.test.js
+++ b/test/logging.test.js
@@ -762,6 +762,44 @@ describe('logging', () => {
     });
   });
 
+  it('should logging in multi-compiler and respect the "stats" option from configuration #3', (done) => {
+    let proc;
+
+    try {
+      proc = execa(runner, [], {
+        stdio: 'pipe',
+        env: {
+          WEBPACK_CONFIG: 'webpack.array.one-error-one-warning-one-object',
+        },
+      });
+    } catch (error) {
+      throw error;
+    }
+
+    let stdout = '';
+    let stderr = '';
+
+    proc.stdout.on('data', (chunk) => {
+      stdout += chunk.toString();
+
+      if (/compiled-for-tests/gi.test(stdout)) {
+        proc.stdin.write('|exit|');
+      }
+    });
+
+    proc.stderr.on('data', (chunk) => {
+      stderr += chunk.toString();
+      proc.stdin.write('|exit|');
+    });
+
+    proc.on('exit', () => {
+      expect(stdoutToSnapshot(stdout)).toMatchSnapshot('stdout');
+      expect(stderrToSnapshot(stderr)).toMatchSnapshot('stderr');
+
+      done();
+    });
+  });
+
   it('should logging an error in "watch" method', (done) => {
     let proc;
 

--- a/test/logging.test.js
+++ b/test/logging.test.js
@@ -186,6 +186,46 @@ describe('logging', () => {
     });
   });
 
+  it("should logging on successfully build when the 'stats' doesn't exist", (done) => {
+    let proc;
+
+    try {
+      proc = execa(runner, [], {
+        stdio: 'pipe',
+        env: {
+          WEBPACK_CONFIG: 'webpack.no-stats.config.js',
+          FORCE_COLOR: true,
+        },
+      });
+    } catch (error) {
+      throw error;
+    }
+
+    let stdout = '';
+    let stderr = '';
+
+    proc.stdout.on('data', (chunk) => {
+      stdout += chunk.toString();
+
+      if (/compiled-for-tests/gi.test(stdout)) {
+        proc.stdin.write('|exit|');
+      }
+    });
+
+    proc.stderr.on('data', (chunk) => {
+      stderr += chunk.toString();
+      proc.stdin.write('|exit|');
+    });
+
+    proc.on('exit', () => {
+      expect(stdout).toContain('\u001b[1m');
+      expect(stdoutToSnapshot(stdout)).toMatchSnapshot('stdout');
+      expect(stderrToSnapshot(stderr)).toMatchSnapshot('stderr');
+
+      done();
+    });
+  });
+
   it('should logging on successfully build and respect the "stats" option from configuration with the "none" value', (done) => {
     let proc;
 
@@ -384,6 +424,7 @@ describe('logging', () => {
         stdio: 'pipe',
         env: {
           WEBPACK_CONFIG: 'webpack.stats-object.config',
+          FORCE_COLOR: true,
         },
       });
     } catch (error) {
@@ -407,6 +448,7 @@ describe('logging', () => {
     });
 
     proc.on('exit', () => {
+      expect(stdout).toContain('\u001b[1m');
       expect(stdoutToSnapshot(stdout)).toMatchSnapshot('stdout');
       expect(stderrToSnapshot(stderr)).toMatchSnapshot('stderr');
 

--- a/test/logging.test.js
+++ b/test/logging.test.js
@@ -725,6 +725,44 @@ describe('logging', () => {
     });
   });
 
+  it('should logging in multi-compiler and respect the "stats" option from configuration #3', (done) => {
+    let proc;
+
+    try {
+      proc = execa(runner, [], {
+        stdio: 'pipe',
+        env: {
+          WEBPACK_CONFIG: 'webpack.array.one-error-one-warning-one-no',
+        },
+      });
+    } catch (error) {
+      throw error;
+    }
+
+    let stdout = '';
+    let stderr = '';
+
+    proc.stdout.on('data', (chunk) => {
+      stdout += chunk.toString();
+
+      if (/compiled-for-tests/gi.test(stdout)) {
+        proc.stdin.write('|exit|');
+      }
+    });
+
+    proc.stderr.on('data', (chunk) => {
+      stderr += chunk.toString();
+      proc.stdin.write('|exit|');
+    });
+
+    proc.on('exit', () => {
+      expect(stdoutToSnapshot(stdout)).toMatchSnapshot('stdout');
+      expect(stderrToSnapshot(stderr)).toMatchSnapshot('stderr');
+
+      done();
+    });
+  });
+
   it('should logging an error in "watch" method', (done) => {
     let proc;
 

--- a/test/logging.test.js
+++ b/test/logging.test.js
@@ -43,10 +43,9 @@ function stdoutToSnapshot(stdout) {
   cleanedStdout = cleanedStdout.replace(/Time: \d+ms/g, 'Time: Xms');
   cleanedStdout = cleanedStdout.replace(/Built at: .+/g, 'Built at: x');
   cleanedStdout = cleanedStdout.replace(/LOG from .+$/s, 'LOG from xxx');
-  cleanedStdout = cleanedStdout.replace(
-    / {3}([a-z]+)\.([a-z]+) +x KiB +\[emitted] +/s,
-    '   $1.$2 x KiB [emitted]'
-  );
+  cleanedStdout = cleanedStdout.replace(/  +/g, ' ');
+  cleanedStdout = cleanedStdout.replace(/^ +/gm, '');
+  cleanedStdout = cleanedStdout.replace(/ +$/gm, '');
 
   return cleanedStdout;
 }

